ast/src/parsed/asm.rs

```diff
@@ -29,26 +29,19 @@ pub struct ASMModule {
     pub statements: Vec<ModuleStatement>,
 }
 
-impl ASMModule {
-    pub fn symbol_definitions(&self) -> impl Iterator<Item = &SymbolDefinition> {
-        self.statements.iter().filter_map(|s| match s {
-            ModuleStatement::SymbolDefinition(d) => Some(d),
-            ModuleStatement::TraitImplementation(_) => None,
-        })
-    }
-}
-
 #[derive(Debug, Clone, PartialEq, Eq, From)]
 pub enum ModuleStatement {
     SymbolDefinition(SymbolDefinition),
-    TraitImplementation(TraitImplementation<Expression>),
+    PilStatement(PilStatement),
 }
 
 impl ModuleStatement {
-    pub fn defined_names(&self) -> Option<&String> {
+    pub fn defined_names(&self) -> Box<dyn Iterator<Item = &String> + '_> {
         match self {
-            ModuleStatement::SymbolDefinition(d) => Some(&d.name),
-            ModuleStatement::TraitImplementation(_) => None,
+            ModuleStatement::SymbolDefinition(d) => Box::new(once(&d.name)),
+            ModuleStatement::PilStatement(s) => {
+                Box::new(s.symbol_definition_names().map(|(name, _)| name))
```

Comment:
I think we should call `symbol_definition_names_and_contained` here, but I'm not sure. `.._and_contained` also returns the variants in an enum for example.

Reply:
So... if we have
```
enum Foo {
    Bar
}
```
Then the defined names should be `[Foo, Bar]`? Or `["Foo", "Foo::Bar"]`?

Reply:
`[("Foo", None), ("Foo", Some("Bar"))]`

Reply:
I mean, what should this function return? Currently it returns all names defined in a given statement, as strings. Should the signature then change?

Reply:
At least before your PR this was essentially only used by the call to local_names() inside check_machine to see if there is a name clash, so I think it's not needed to return the nested names.

Reply:
It's used only in the importer to prevent duplicates, and to detect the standard library. I think we do not need the nested names for that. Then the current code is correct?